snuba/clickhouse/native.py

```diff
@@ -219,8 +219,12 @@ def __transform_result(self, result, with_totals: bool) -> Result:
 
     def execute(
         self,
-        query: ClickhouseQuery,
-        # TODO: move Clickhouse specific arguments into DictClickhouseQuery
+        # TODO: After we remove DictSqlQuery (that does all query processing for the
```

Comment:
I tried a few alternatives here:
- just pass the Clickhosue Query. Does not work because other parts of the query processor before depend on the formatted query and we cannot format the old query representation twice (the AST solves this) since the formatting process changes the original query (not kidding).
- call the SqlQuery a "formatter". That would feel confusing since we would pass a "formatter" and not a query to the reader.
- frame the SqlQuery as a wrapper for the Clickhosue Query that provides formatting.  This is where I landed.

I think we can live with this till the AST is done (specifically now that we are actively working towards that) so that we can treat formatting as the stateless process it is supposed to be, let the Reader call the formatter and remove all the dependencies on the formatted query before the Reader.

Reply:
Thanks for the explanation, this is really helpful.